### PR TITLE
Add Thrift Data Publisher & Data Provider

### PR DESCRIPTION
**Pull request by senthuran16**
# Purpose
> $subject.

API Analytics events from WSO2 APIM 4.2.0 are published to the following Thrift streams:
- Request Events: `org.wso2.apimgt.statistics.request:4.2.0`
- Fault Events: `org.wso2.apimgt.statistics.fault:4.2.0`
- Throttle Events: `org.wso2.apimgt.statistics.throttle:4.2.0`
See the topic **Thrift stream Attributes** for the list of attributes of the above streams.

### Note: Fault Events
In WSO2 APIM 4.2.0, Some error events might be reported with `null` attributes (eg: `applicationConsumerKey` will be `null` when invoking an API with an invalid token), and fail during validation - since required attributes for the fault Thrift stream are null.

During such cases, the validation fails and these events are not reported. The error that is thrown during validation will NOT be logged by default - to avoid the log growth during too many such erroneous calls. If you need to see the error logs upon validation failure, add the following to the `<WSO2_APIM_4.2.0_HOME>/repository/conf/log4j2.properties` file:
```
loggers = thrift-parallel-queue-worker # ...rest of the loggers

logger.thrift-parallel-queue-worker.name = org.wso2.am.thrift.analytics.publisher.ParallelQueueWorker
logger.thrift-parallel-queue-worker.level = DEBUG
```

# Goals
> Publishing API analytics events from WSO2 APIM 4.x to a Thrift server.

# Instructions

## Building the artifacts (Thrift Data Publisher & Data Provider).

Run `mvn clean install` from the respective directory (`thrift-analytics-data-provider` or `thrift-analytics-publisher`)

## Configurations to Enable Thrift Reporting
Go to `<WSO2_APIM_4.2.0_HOME>/repository/conf/deployment.toml`, and provide the following configurations to enable Thrift analytics publishing:
```toml
[apim.analytics]
enable = true
properties."publisher.custom.data.provider.class" = "org.wso2.am.thrift.analytics.data.provider.ThriftAnalyticsDataProvider"
properties."publisher.reporter.class" = "org.wso2.am.thrift.analytics.publisher.ThriftMetricReporter"
properties."thrift.endpoint.agent.type" = "THRIFT"
properties."thrift.endpoint.receiver.url.set" = "tcp://0.0.0.0:7612" # Point this to a Thrift server
properties."thrift.endpoint.auth.url.set" = "ssl://0.0.0.0:7712" # Point this to a Thrift Server
properties."thrift.endpoint.username" = "admin"
properties."thrift.endpoint.password" = "admin"
properties.build_response_message = true # Enable this to collect the response message size

# Optional
properties."thrift.data.publisher.event.queue.size" = 32768 # Size of the event queue 
properties."thrift.data.publisher.worker.thread.count" = 10 # No. of worker threads that will take events from the event queue
properties."thrift.data.publisher.pool.max.idle" = 250
properties."thrift.data.publisher.pool.init.idle.capacity" = 200

```

## Logger Configurations for the Custom Thrift Data Provider & Publisher
Go to the `loggers` section of the `<WSO2_APIM_4.2.0_HOME>/repository/conf/log4j2.properties` file, and provide the following:

```
loggers = thrift-analytics-publisher, thrift-analytics-data-provider, # ...rest of the loggers

# Custom Thrift Publisher
logger.thrift-analytics-publisher.name = org.wso2.am.thrift.analytics.publisher
logger.thrift-analytics-publisher.level = INFO

# Thrift Data Provider
logger.thrift-analytics-data-provider.name = org.wso2.am.thrift.analytics.data.provider
logger.thrift-analytics-data-provider.level = INFO
```

# Thrift stream Attributes
### Request Events: `org.wso2.apimgt.statistics.request:4.2.0`
Attribute | Type | Description
-- | -- | --
meta_clientType | string | The meta information of client type.
applicationConsumerKey | string | The consumer key of the API invoked client application.
applicationName | string | The name of the client application.
applicationId | string | The ID of the client application.
applicationOwner | string | The name of the owner of the application.
apiContext | string | The API context depending on the user's request.
apiName | string | The API name.
apiVersion | string | The API version.
apiResourcePath | string | The API resource path of the API request.
apiResourceTemplate | string | The API resource URL pattern of the API request.
apiMethod | string | The HTTP verb of the API request [e.g.,GET/POST].
apiCreator | string | The creator of the API.
apiCreatorTenantDomain | string | The tenant domain of the API creator.
apiTier | string | The subscription tier associated with the API request.
apiHostname | string | The hostname or Datacenter ID (if specified).
username | string | The end-user of the API request.
userTenantDomain | string | The tenant domain of the user that is associated with the request.
userIp | string | The IP address of the client.
userAgent | string | The user agent of the user.
requestTimestamp | long | The timestamp of the API request when received at the Gateway.
throttledOut | bool | This describes whether the request was allowed after hitting the throttle tier.
responseTime | long | The timestamp of the API response when received at the Gateway.
serviceTime | long | The time that is taken to serve the API request at the API-M side.
backendTime | long | The time taken to process the request at the backend.
responseCacheHit | bool | This describes if response caching is enabled or not.
responseSize | long | The response message size in bytes.
protocol | string | The protocol used to send the response (HTTP/HTTPS) and the port.
responseCode | int | The HTTP response code.
destination | string | The URL of the endpoint.
securityLatency | long | The time taken for authentication.
throttlingLatency | long | The time taken for throttling the request/response.
requestMedLat | long | The time taken to mediate the request.
responseMedLat | long | The time taken to mediate the response.
backendLatency | long | The time taken by the backend to return the response.
otherLatency | long | The time taken to process tasks other than mentioned above.
gatewayType | string | The Gateway type (Synapse/Micro).
label | string | The label of the API (if specified).
properties | string | The JSON string with custom attributes (if specified).

### Fault Events: `org.wso2.apimgt.statistics.fault:4.2.0`
Attribute | Type | Description
-- | -- | --
meta_clientType | string | The meta information of client type.
applicationConsumerKey | string | The consumer key of the API invoked client application.
apiName | string | The API name.
apiVersion | string | The API version.
apiContext | string | The API context depending on the user's request.
apiResourcePath | string | The API resource path of the API request.
apiResourceTemplate | string | The API resource URL pattern of the API request.
apiMethod | string | The HTTP verb of API request [e.g.,GET/POST].
apiCreator | string | The creator of the API.
username | string | The end-user of the API request.
userTenantDomain | string | The tenant domain of the user that is associated with the request.
apiCreatorTenantDomain | string | The tenant domain of the API creator.
hostname | string | The hostname or datacenter ID (if specified).
applicationId | string | The ID of the client application.
applicationName | string | The name of the client application.
applicationOwner | string | The name of the owner of the application.
protocol | string | The protocol used to send the response (HTTP/HTTPS) and the port.
errorCode | string | The Synapse error code.
errorMessage | string | The description of the Synapse error message.
requestTimestamp | long | The timestamp of the API request when received at the Gateway.
properties | string | The JSON string with custom attributes (if specified).

### Throttle Events: `org.wso2.apimgt.statistics.throttle:4.2.0`
Attribute | Type | Description
-- | -- | --
meta_clientType | string | The meta information of client type.
username | string | The end-user of the API request.
userTenantDomain | string | The tenant domain of the user that is associated with the request.
apiName | string | The API name.
apiVersion | string | The APIversion.
apiContext | string | The API context depending on the user's request.
apiCreator | string | The creator of the API.
apiCreatorTenantDomain | string | The tenant domain of the API creator.
apiResourceTemplate | string | The API resource URL pattern of the API request.
apiMethod | string | The HTTP verb of API request [e.g.,GET/POST].
applicationId | string | The ID of the client application.
applicationName | string | The name of the client application.
subscriber | string | The name of the subscriber of the application.
throttledOutReason | string | The reason describing why the request has been throttled out.
gatewayType | string | The Gateway type (Synapse/Micro).
throttledOutTimestamp | long | The timestamp when the request is throttled out.
hostname | string | The hostname or datacenter ID (if specified).
properties | string | The JSON string with custom attributes (if specified).

### CODE REVIEW

File: thrift-analytics-publisher/pom.xml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.am.thrift.analytics.publisher.*;version="${project.version}"
+                        </Export-Package>
+                        <Private-Package>
+                        </Private-Package>
+                        <Import-Package>
+                            org.apache.synapse.*,
+                            org.apache.commons.logging.*,
+                            org.wso2.carbon.apimgt.common.analytics.*;version=${carbon.apimgt.version},
```
Comment (AnuGayan) [inline thrift-analytics-publisher/pom.xml:93]: Shall we import a range here,

File: thrift-analytics-publisher/src/main/java/org/wso2/am/thrift/analytics/publisher/ThriftClient.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        Event event = new Event();
+        String streamId = DataBridgeCommonsUtils.generateStreamId(streamName, APIM_420_STREAM_VERSION);
+        event.setStreamId(streamId);
+        event.setMetaData(null);
+        event.setCorrelationData(null);
+        event.setPayloadData(payload);
+
+        try {
+//            printEventMap(eventMap); // Can be used to print the event for debugging purposes
```
Comment (AnuGayan) [inline thrift-analytics-publisher/src/main/java/org/wso2/am/thrift/analytics/publisher/ThriftClient.java:82]: Let's remove this, and Improve and add some debug logs throughout the class

Comment (senthuran16) [reply thrift-analytics-publisher/src/main/java/org/wso2/am/thrift/analytics/publisher/ThriftClient.java]: Fixed via https://github.com/wso2/samples-apim/pull/119/commits/9ba77338d22d6c68238a1285e7ecfad7d0cd11c6

File: thrift-analytics-publisher/pom.xml (hunk 1 above)
Comment (senthuran16) [reply thrift-analytics-publisher/pom.xml]: Addressed via https://github.com/wso2/samples-apim/pull/119/commits/9ba77338d22d6c68238a1285e7ecfad7d0cd11c6

File: thrift-analytics-data-provider/src/main/java/org/wso2/am/thrift/analytics/data/provider/ThriftAnalyticsDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+    private String getApiHostName(MessageContext messageContext) {
+        return GatewayUtils.getHostName(messageContext);
+    }
+
+    private String getUserTenantDomain(AuthenticationContext authenticationContext) {
+        if (authenticationContext != null) {
+            return MultitenantUtils.getTenantDomain(authenticationContext.getUsername());
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (senthuran16) [reply thrift-analytics-data-provider/src/main/java/org/wso2/am/thrift/analytics/data/provider/ThriftAnalyticsDataProvider.java]: Addressed via https://github.com/wso2/samples-apim/pull/119/commits/f7f94219c7becf73511b7f92a5130aff75040644

**Review by AnuGayan — APPROVED** [review]

**Pull request merged**